File: Top_100_Liked_Questions/160._Intersection_of_Two_Linked_Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None
        
        p1 = headA
        p2 = headB
        
        cnt = 0
        while cnt <= 2:
            if p1 is p2:
                return p1
            
            if p1.next:
                p1 = p1.next
            else:
                p1 = headB
                cnt += 1
            
            if p2.next:
                p2 = p2.next
            else:
                p2 = headA
                cnt += 1
        
        return None

File: Top_100_Liked_Questions/160._Intersection_of_Two_Linked_Lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null)
            return null;

        ListNode pt1 = headA;
        ListNode pt2 = headB;

        int flipCount = 0;

        while (flipCount <= 2) {
            if (pt1 == pt2)
                return pt1;

            if (pt1.next != null)
                pt1 = pt1.next;
            else {
                pt1 = headB;
                flipCount++;
            }

            if (pt2.next != null)
                pt2 = pt2.next;
            else {
                pt2 = headA;
                flipCount++;
            }
        }

        return null;
    }
}